extensions/dim-rename-kafka8/src/main/java/io/druid/server/namespace/KafkaExtractionManager.java

```diff
@@ -0,0 +1,231 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.namespace;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Key;
+import com.google.inject.name.Names;
+import com.metamx.common.IAE;
+import com.metamx.common.ISE;
+import com.metamx.common.StringUtils;
+import com.metamx.common.lifecycle.LifecycleStart;
+import com.metamx.common.lifecycle.LifecycleStop;
+import com.metamx.common.logger.Logger;
+import io.druid.guice.ManageLifecycle;
+import io.druid.query.extraction.namespace.KafkaExtractionNamespace;
+import kafka.consumer.ConsumerConfig;
+import kafka.consumer.ConsumerIterator;
+import kafka.consumer.KafkaStream;
+import kafka.consumer.Whitelist;
+import kafka.javaapi.consumer.ConsumerConnector;
+import kafka.message.MessageAndMetadata;
+import kafka.serializer.Decoder;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.regex.Pattern;
+
+/**
+ *
+ */
+@ManageLifecycle
+public class KafkaExtractionManager
+{
+  private static final Logger log = new Logger(KafkaExtractionManager.class);
+
+  private final Properties kafkaProperties;
+  private final Collection<ListenableFuture<?>> futures = new ConcurrentLinkedQueue<>();
+  private final ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(
+          new ThreadFactoryBuilder()
+              .setNameFormat("kafka-rename-consumer-%d")
+              .setDaemon(true)
+              .setPriority(Thread.MIN_PRIORITY)
+              .build()
+      )
+  );
+  private final AtomicInteger startedPipes = new AtomicInteger(0);
+  private final AtomicLong events = new AtomicLong(0l);
+
+  // Bindings in KafkaExtractionNamespaceModule
+  @Inject
+  public KafkaExtractionManager(
+      Injector injector
```

Comment:
One of the primary benefits of DI is that you get objects which are defined in terms of their dependencies.  This constructor is making the statement that a `KafkaExtractionManager` depends on the entire global state of the process (i.e., the `Injector`), which is incorrect.

The proper way to work with DI is to define the object depending directly on its dependencies and then setup your bindings so that the proper things get sent in.

In this case, I think what you want is to create a `KafkaExtractionManagerConfig` object that presently just has a single `Map<String, String> kafkaProperties` field.  I would then create a static `defaultsMap` object in that class and have it actually store the composition of the `defaultsMap` with the provided `kafkaProperties`.  If you structure it so that `kafkaProperties` keys overwrite `defaultsMap` keys in the local map, then you get a lot cleaner implementation of defaulting.  After doing that, you can do the test for bad keys and you are good.

On the "don't do anything if namespacing is disabled" case.  This object shouldn't even exist if namespacing is disabled.  The bindings and object graph should guarantee that, so please rework how the bindings work such that this object will never be instantiated if it shouldn't do anything.